### ISSUE TITLE
Why was ChoiceType "choices" option convention changed?

### ISSUE THREAD

**Issue by justin-oh** [symfony/symfony#19953]
I've done several Symfony projects beyond version 2.7.* in which I've used the `ChoiceType ` form element by setting the `choices` option to an array using the convention `value => label`.

However, I just noticed the convention has changed to `label => value`, via http://symfony.com/blog/new-in-symfony-2-7-choice-form-type-refactorization and http://symfony.com/doc/current/reference/forms/types/choice.html#choices.

This change doesn't make sense to me. The convention has always been `value => label` anytime I've worked with a form APIs. This convention is intuitive with the uniqueness of array keys. It also mimics the `<select>` element.

Flipping the convention, I see at least three issues:
- potential to have duplicate values in a form selection (I can't think of a reason this is useful)
- potential to lose a selection if two labels are equal (eg. a User selection where their names are labels)
- everyone has to change their code and way of thinking to move forward with Symfony

Can someone elaborate as to why this drastic of a change needed to happen? If the `choices` **really** needed to be backwards, that could have been done internally without affecting convention.

**Comment by justin-oh** [comment]
I know "choices_as_values" is there for backwards compatibility in +2.7 and is removed in 3.0. I think it should be re-added but as an app-wide config option, so developers can continue to use their preferred convention. Example:

```yaml
# app/config.yml
framework:
    form:
        choices_as_values: false # defaults to true
```

**Comment by iltar** [comment]
This setup allows the following:
 - Label: Foo; value: 1
 - Label: Fooo; value: 1
 - Label: Bar; value 2

Not sure what the actual reason was, but I don't really mind either way.

**Comment by xabbuh** [comment]
Flipping the choices makes it possible to use other types than strings or integers as choices (PHP doesn't allow other types as array keys and the label is a string anyway). See #14050 for more details.

**Comment by justin-oh** [comment]
I guess the question I wanted answered from the beginning is: why are we being forced to use the new convention in 3.0? Why not leave it as an option to those that wish to pass in objects for `choices`?

**Comment by naitsirch** [comment]
@xabbuh But in HTML option values are only strings, too. What is the advantage in having other types for labels in PHP?

**Comment by xabbuh** [comment]
@naitsirch You don't want to have other types for the label. That's why it is precisely the key of the array. However, the choice itself may be any kind of PHP data type.

**Comment by justin-oh** [comment]
@xabbuh If someone is passing in other types of data, it should be a class object. It would be convention to have the class object implement __toString for the label, and the choices convention can remain unchanged.

**Comment by xabbuh** [comment]
The values don't have to be objects. It could also be simple boolean values, for example.

**Comment by justin-oh** [comment]
Yes, but PHP/Symfony users have been getting alone fine under the current guidelines for array keys. The ability to use simple data (that is not strings or integers) as the "value" is just a very minor convenience. It's not an issue enough in itself to flip a whole convention.

The real reason the convention was flipped in Symfony was to accommodate for objects as the "value". I believe it was a poor design choice because it adds another step to upgrading to Symfony 3. The developers should want that upgrade process to be as seamless as possible.

What should have been developed was an intermediate field between ChoiceType and EntityType (perhaps called ObjectType). This ObjectType would expect an array of class objects as the "choices", and the convention would be that these classes implement specific functions for retrieving the default value/label. (The Form code could even provide an interface for this purpose.) This way, the convention doesn't change, and ChoiceType doesn't become more complicated than it needs to be.

**Comment by dbrumann** [comment]
Even with values that are booleans or null, when you use them as keys they will automatically be transformed, see: https://stackoverflow.com/a/45417057/1166880
You could also create weird edge cases with floats and ints like this https://3v4l.org/CIdZi.

For this reason it's safer for the labels to be used as keys, as they will always be strings no matter what and you can make sure that the data in the values is not magically transformed by PHP because of it's limitations on what arrays keys should look like.

**Comment by justin-oh** [comment]
I know that happens. My point is it's been happening for quite a few years and nothing has been done about it. This leads me to believe it's nothing more than a minor inconvenience, and not reason enough alone to flip the convention. It should've at least been left in as an option for those that wish to flip it.

**Comment by stof** [comment]
@justin-oh during years, using boolean values was impossible when using `choices`, and required using the lower-level `choice_list` API directly. And people reported us bugs about boolean or float choices being broken at least 20 times over years before we made the flip.

Btw, it is weird to complain about it now. This flip was done in 2.8, which means 2 years ago. Why resurrecting this discussion ?

**Comment by naitsirch** [comment]
Sorry, I didn't want to complain. It's just because I haven't got the point, yet.

> @naitsirch You don't want to have other types for the label. That's why it is precisely the key of the array. However, the choice itself may be any kind of PHP data type.

@xabbuh Sorry, I mixed it up. I wanted to say "What is the advantage in having other types for values in PHP?".

But now after trying it out with a small example I realized, what Symfony does and why.

```php
$builder->add('active', ChoiceType::class, ['choices' => [
    'No' => false,
    'Yes' => true,
    'Maybe' => 0,
]]);
```
This will create an HTML select field like this:
```html
<select id="form_active" name="form[active]">
    <option value="0">No</option>
    <option value="1">Yes</option>
    <option value="2">Maybe</option>
</select>
```
It wasn't clear to me, that Symfony will use incrementing integers as values, if there are other types than strings passed as values and use the correct value with the correct type when mapping the form to the object.

But as far as I see, this behavior is not documented in the [ChoiceType](https://symfony.com/doc/4.0/reference/forms/types/choice.html#choices), or am I missing something?
